sys/include/net/sock/dtls.h

```diff
@@ -0,0 +1,557 @@
+ * endpoint using @ref sock_dtls_establish_session(). If the handshake is
+ * successfull and the session is established, we can use its `remote` to send
+ * packets to it with @ref sock_dtls_send().
+ *
+ * If error during any of these operation, the session is closed and the sock
+ * is destroyed using @ref sock_dtls_close_session() and
+ * @ref sock_dtls_destroy().
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * res = sock_dtls_create(&dtls_sock, &udp_sock, DTLS_SOCK_CLIENT_TAG, 0);
+ * if (res < 0) {
+ *     puts("Error creating DTLS sock");
+ *     return;
+ * }
+ *
+ * res = sock_dtls_establish_session(&dtls_sock, &remote, &session);
+ * if (res < 0) {
+ *     printf("Error establishing session: %d\n", (int)res);
+ *     goto end;
+ * }
+ *
+ * res = sock_dtls_send(&dtls_sock, &session, data, datalen);
+ * if (res < 0) {
+ *     printf("Error sending DTLS message: %d\n", (int)res);
+ *     goto end;
+ * }
+ *
+ * res = sock_dtls_recv(&dtls_sock, &session, rcv, sizeof(rcv), SOCK_NO_TIMEOUT);
+ * if (res < 0) {
+ *     printf("Error receiving DTLS message: %d\n", (int)res);
+ *     goto end;
+ * }
+ *
+ * end:
+ * puts("Terminating");
+ * sock_dtls_close_session(&dtls_sock, &session);
+ * sock_dtls_destroy(&dtls_sock);
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * @{
+ *
+ * @file
+ * @brief   DTLS sock definitions
+ *
+ * @author  Aiman Ismail <muhammadaimanbin.ismail@haw-hamburg.de>
+ */
+
+#ifndef NET_SOCK_DTLS_H
+#define NET_SOCK_DTLS_H
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/types.h>
+
+#include "net/sock.h"
+#include "net/sock/udp.h"
+#include "net/credman.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Type for a DTLS sock object
+ *
+ * @note    API implementors: `struct sock_dtls` needs to be defined by
+ *          implementation-specific `sock_dtls_types.h`.
+ */
+typedef struct sock_dtls sock_dtls_t;
+
+/**
+ * @brief Information about the established session with
+ *        the remote endpoint. Used when sending and
+ *        receiving data to the endpoint
+ */
+typedef struct sock_dtls_session sock_dtls_session_t;
+
+/**
+ * @brief Must be called once before any other use
+ */
+void sock_dtls_init(void);
+
+/**
+ * @brief Creates a new DTLS sock object
+ *
+ * @param[out] sock     The resulting DTLS sock object
+ * @param[in] udp_sock  Existing UDP sock to be used
+ * @param[in] tag       Credential tag of the sock. Used to get the right
+ *                      credential from pool.
+ * @param[in] method    Defines the method for the client or server to use.
+ *
+ * @return  0 on success.
+ * @return  value < 0 on error
+ */
+int sock_dtls_create(sock_dtls_t *sock, sock_udp_t *udp_sock,
+                     credman_tag_t tag,unsigned method);
+
+/**
+ * @brief Initialises the server to listen for incoming connections
+ *
+ * @param[in] sock      DTLS sock to listen to
+ */
+void sock_dtls_init_server(sock_dtls_t *sock);
```

Comment:
In accordance with `gnrc_tcp`: maybe name it `gnrc_dtls_listen()`? Not sure, if we already talked about this :sweat_smile:...

Reply:
The description might be misleading. There is actually no 'listening' that happens here. What I meant to say is, the function initializes the sock to for any configurations that must be done so that it can be used as a DTLS server.  So maybe:

```C
/**
 * @brief configure the @p sock to be used as a DTLS server.
 *
 * @param[in] sock      DTLS sock
 */
void sock_dtls_init_server(sock_dtls_t *sock);
```

DTLS is actually more like UDP than TCP. There is no specific "listening for new connection/session" function. The listening is done by the receiving function e.g. 'sock_dtls_recv()`. Handshake process are started when a DTLS server received a ClientHello message instead of Application Data from a client.

Reply:
Ok.